### PR TITLE
Shared: restrict flow after using implicit read

### PR DESCRIPTION
**Pull request by asgerf**
After using an implicit read, flow can now only terminate in a sink or continue along edges contributed by `Config:isAdditionalFlowStep`, but not via other kinds of steps, like simple local flow steps.

This is achieved by treating the implicit read node as a sink if its original node was a sink, and copying any relevant outgoing steps from the original node onto the implicit read node.

### CODE REVIEW

**Review by aschackmull — APPROVED** [review]
Looks correct to me.

**Comment by aschackmull** [comment]
Are the dca runs up-to-date and ok? If so, then this looks ready to merge.

**Comment by asgerf** [comment]
Yes, the DCA runs are up to date. (The C++ report is messed up due to a failure in JavaScriptCore, but that's also failing on nightly)

**Comment by jketema** [comment]
This will need to be backed out. The C++ DCA failure on Wireshark is a serious regression where we now get a timeout.

File: cpp/ql/test/library-tests/dataflow/taint-tests/taint.expected [hunk 1]
```diff
@@ -4,4 +4,7 @@ WARNING: module 'DataFlow' has been deprecated and may be removed in future (tai
 WARNING: module 'DataFlow' has been deprecated and may be removed in future (taint.ql:68,25-33)
 WARNING: module 'TaintTracking' has been deprecated and may be removed in future (taint.ql:73,20-33)
 testFailures
+| taint.cpp:453:23:453:42 | // $ ir MISSING: ast | Missing result:ir= |
+| vector.cpp:118:12:118:30 | // $ ir MISSING:ast | Missing result:ir= |
+| vector.cpp:119:12:119:30 | // $ ir MISSING:ast | Missing result:ir= |
```
Comment (geoffw0) [inline cpp/ql/test/library-tests/dataflow/taint-tests/taint.expected:9]: Inline expectations tests should not output anything.  The annotations should be updated if we accept these changes.

Comment (jketema) [reply cpp/ql/test/library-tests/dataflow/taint-tests/taint.expected:9]: This was fixed in https://github.com/github/codeql/pull/17275, I believe.

Comment (geoffw0) [reply cpp/ql/test/library-tests/dataflow/taint-tests/taint.expected:9]: Excellent, thanks.

**Pull request merged**